### PR TITLE
Token spawn refactoring

### PR DESCRIPTION
**Pull request by Buhallin**
This ended up unfortunately large, but once it was in flight I couldn't find a decent place to split it that wouldn't leave something broken.
- Factors out token spawning responsibilities to a module
- Spawns tokens by pulling from the appropriate infinite bag, rather than creating from scratch
--- This required flipping the doom containers and the tokens in them to ensure they were programatically the right side up
- Adds ability for all cards to spawn based on uses metadata, DataHelper data (including Custom Data Helpers) will be used as a fallback if metadata is missing
- Adds a new dedicated object to track token spawn state for cards (previously done in DataHelper, moved for single responsibility principle)
- Filters which cards will spawn tokens based on the play area:
--- Playmat slots area will only spawn for assets and events
--- Playmat threat areas will spawn for enemies, treacheries, and weaknesses
--- Main play area spawn for locations, enemies, treacheries, and weaknesses
- Adds listeners to all typical decks, discard piles, and hands to reset a card's token spawn when it leaves play

### CODE REVIEW

File: src/playermat/Playmat.ttslua [hunk 1]
```diff
@@ -1,5 +1,7 @@
+local tokenManager = require("core/token/TokenManager")
+
 -- set true to enable debug logging and show Physics.cast()
-local DEBUG = false
+local DEBUG = true
```
Comment (Chr1Z93) [inline src/playermat/Playmat.ttslua:4]: This is not intended, is it?

File: src/playermat/Playmat.ttslua [hunk 2]
```diff
@@ -457,133 +434,102 @@ function replenishTokens(card, count, replenish)
 
   local newCount = foundTokens + replenish
   if newCount > count then newCount = count end
-  spawnTokenGroup(card, "resource", newCount)
-end
-
-function getPlayerCardData(object)
-  return PLAYER_CARDS[object.getName()..':'..object.getDescription()] or PLAYER_CARDS[object.getName()]
+  tokenManager.spawnTokenGroup(card, "resource", newCount)
```
Comment (Chr1Z93) [inline src/playermat/Playmat.ttslua:437]: Looks like this is part of the replenish function? Does not look intended to me since it doesn't take the replenish count as argument

File: src/playermat/Playmat.ttslua [hunk 3]
```diff
@@ -457,133 +434,102 @@ function replenishTokens(card, count, replenish)
...
+    for _, obj in ipairs(searchArea(PLAY_ZONE_POSITION, PLAY_ZONE_SCALE)) do
+      local obj = obj.hit_object
+      local notes = JSON.decode(obj.getGMNotes()) or { }
+      if notes.id == (cardMetadata.id .. "-c") then
+        for i, customization in ipairs(cardMetadata.customizations) do
+          if obj.getVar("markedBoxes")[i] == customization.xp
+              and customization.replaces ~= nil
+              and customization.replaces.uses ~= nil then
+            cardMetadata.uses = customization.replaces.uses
+            card.setGMNotes(JSON.encode(cardMetadata))
```
Comment (Chr1Z93) [inline src/playermat/Playmat.ttslua:452]: How does this work with cards like `Refine` that add checkboxes during play?

File: src/playermat/Playmat.ttslua [hunk 4]
```diff
@@ -680,7 +622,7 @@ function showDrawButton(visible)
   -- remove the "Draw 1" button
   else
     local buttons = self.getButtons()
-    for i = 1, #buttons do 
+    for i = 1, #buttons do
```
Comment (Chr1Z93) [inline src/playermat/Playmat.ttslua:625]: Is this a LF related change?

File: src/playermat/Playmat.ttslua [hunk 5]
```diff
@@ -721,7 +663,7 @@ function clickableClues(showCounter)
     local pos = self.positionToWorld({x = -1.12, y = 0.05, z = 0.7})
     for i = 1, clueCount do
       pos.y = pos.y + 0.045 * i
-      spawnToken(pos, "clue")
+      TokenManager.spawnToken(pos, "clue", self.getRotation())
```
Comment (Chr1Z93) [inline src/playermat/Playmat.ttslua:666]: Playermat rotation is already stored in a variable

**Review by Chr1Z93 — COMMENTED** [review]
First batch of comments added (will take deeper look when I am not on mobile)

File: objects/LuaScriptState.luascriptstate [hunk 6]
```diff
@@ -1 +1 @@
-{"optionPanel":{"showChaosBagManager":false,"showCleanUpHelper":false,"showDrawButton":false,"showHandHelper":{},"showNavigationOverlay":false,"showTokenArranger":false,"useClueClickers":false,"useSnapTags":true}}
+{"optionPanel":{"showChaosBagManager":false,"showCleanUpHelper":false,"showDrawButton":false,"showHandHelper":[],"showNavigationOverlay":false,"showTokenArranger":false,"useClueClickers":false,"useSnapTags":true}}
```
[the remark this answers is not in the document]
Comment (Buhallin) [reply objects/LuaScriptState.luascriptstate:1]: Err, no idea honestly.  Will fix directly and look into what happened.

[the remark this answers is not in the document]
Comment (Buhallin) [reply objects/LuaScriptState.luascriptstate:1]: Ah, I see - I think this is because the hand helper option panel is an array now, not a single value.  Not sure why that didn't commit before, but it should be correct.

File: src/arkhamdb/DeckImporterMain.ttslua [hunk 7]
```diff
@@ -202,8 +202,7 @@ end
 -- Returns the simple name of a card given its ID.  This will find the card and strip any trailing
 -- SCED-specific suffixes such as (Taboo) or (Level)
 function getCardName(cardId)
-  local configuration = getConfiguration()
-  local allCardsBag = getObjectFromGUID(configuration.card_bag_guid)
+  local allCardsBag = getObjectFromGUID("15bb07")
```
[the remark this answers is not in the document]
Comment (Buhallin) [reply src/arkhamdb/DeckImporterMain.ttslua:205]: It actually does need to be here :)  Since the refactoring, DeckImporterMain doesn't have a reference to the configuration token any more.  It only matters for this one use (getting names for Raven Quill) but this was broken.

I didn't make it a constant since it's a single, well-known use that's obvious from the context, and one of my to-dos is to extract an API for AllCardsBag which would replace it.

File: src/core/Global.ttslua [hunk 8]
```diff
@@ -614,7 +595,7 @@ end
 -- Content Importing and XML functions
 ---------------------------------------------------------
 
-local source_repo = 'https://raw.githubusercontent.com/seth-sced/loadable-objects/main'
+local source_repo = 'https://raw.githubusercontent.com/chr1z93/loadable-objects/main'
```
[the remark this answers is not in the document]
Comment (Buhallin) [reply src/core/Global.ttslua:598]: I'm torn on this.  Generally I'd agree, but there's enough in here that relies on the metadata that we're going to need to point at the new ones for a LOT of testing, including the beta release.  We can leave the prod version here and change it in the release branch, or we can commit this and change it back to prod once the new metadata versions are live.

I'm inclined to cheat on this one and point everything to this version for now, but I'm happy to bow to being more strict about it.  Going to leave this open pending further comment.

File: src/core/Global.ttslua [hunk 9]
```diff
@@ -847,7 +828,7 @@ function applyOptionPanelChange(id, state)
   -- option: Show clean up helper
   elseif id == "showCleanUpHelper" then
     optionPanel[id] = spawnOrRemoveHelper(state, "Clean Up Helper", {-68, 1.6, 35.5})
-  
+
```
[the remark this answers is not in the document]
Comment (Buhallin) [reply src/core/Global.ttslua:831]: I don't use autoformat :)  I don't remember doing this but I probably stumbled across the trailing whitespace and deleted it manually.

File: src/core/MythosArea.ttslua [hunk 10]
```diff
@@ -1,4 +1,14 @@
...
+local tokenSpawnTracker = require("core/token/TokenSpawnTrackerApi")
+
+local ENCOUNTER_DECK_AREA = {
+  upperLeft = { x = 0.9, z = 0.42 },
+  lowerRight = { x = 0.86, z = 0.38 },
+}
+local ENCOUNTER_DISCARD_AREA = {
+  upperLeft = { x = 1.62, z = 0.42 },
+  lowerRight = { x = 1.58, z = 0.38 },
+}
```
[the remark this answers is not in the document]
Comment (Buhallin) [reply src/core/MythosArea.ttslua:11]: Actually, the ones in Global should be removed in favor of these :)  These decks belong to the Mythos area, and the draws should be its responsibility.  We can replace the Global access to them with an API on the mythos area to at least get the decks, and possibly handle the draw.

At the very least these are local, while the Global is, uhm, global... which makes the Global entries fragile if we move this panel.

File: src/playermat/Playmat.ttslua (hunk 1 above)
Comment (Buhallin) [reply src/playermat/Playmat.ttslua:4]: Definitely not.  Fixed.

File: src/playermat/Playmat.ttslua (hunk 2 above)
Comment (Buhallin) [reply src/playermat/Playmat.ttslua:437]: The diff on this is misleading.  It's replacing this line:
https://github.com/argonui/SCED/blob/ff7ed57bca58989372eb003193624f059f63b0e2/src/playermat/Playmat.ttslua#L460

It's identical except for using the TokenManager to do the spawning instead of the local method.

File: src/playermat/Playmat.ttslua (hunk 4 above)
Comment (Buhallin) [reply src/playermat/Playmat.ttslua:625]: I'm not sure honestly - doesn't look like it, most likely another trailing space I stumbled across and cleaned up.

File: src/playermat/Playmat.ttslua (hunk 5 above)
Comment (Buhallin) [reply src/playermat/Playmat.ttslua:666]: Changed to be consistent(ish, the file isn't consistent) but I'm not sure this is a good approach.  If the mat were to rotate, this would have an incorrect value.  Pretty unlikely with the playmats, but as a general rule I think anything about object state should be retrieved directly when needed at runtime rather than cached like this

File: src/playermat/Playmat.ttslua (hunk 3 above)
Comment (Buhallin) [reply src/playermat/Playmat.ttslua:452]: At the moment it doesn't.  We need to add something to the Upgrade Sheets to push a call to this for all cards on the mat.  I think we can do that in a followon.

But I also don't think it affects anything.  For everything except Runic Axe, the only changes are to the initial number of uses.  So a change won't affect anything already in play.  The only issue would be Runic Axe's extra replenishment, which is still handled explicitly.

File: src/core/PlayArea.ttslua [hunk 11]
```diff
@@ -64,84 +49,35 @@ function log(message)
...
   if not COLLISION_ENABLED then return end
 
   -- check if we should spawn clues here and do so according to playercount
-  local object = collision_info.collision_object
-  if getLocation(object) ~= nil and spawnedLocationGUIDs[object.getGUID()] == nil then
-    local clueCount = getClueCount(object, object.is_face_down, getInvestigatorCount())
-    if clueCount > 0 then spawnCluesAtLocation(clueCount, object) end
+  local card = collision_info.collision_object
+  if shouldSpawnTokens(card) then
+    tokenManager.spawnForCard(card)
```
[the remark this answers is not in the document]
Comment (Buhallin) [reply src/core/PlayArea.ttslua:68]: It could, but it would add some complexity that I don't think is necessary.  The JSON parser may be slow, but our metadata is still small and doing it for a specific card shouldn't add any appreciable time.  Even the worst case for this (placing a scenario that drops as many as 20) will be overwhelmed by the spawn time in TTS.

We should definitely keep an eye on this and see if it starts causing problems, but until it does I consider it a premature optimization.

File: src/core/token/TokenManager.ttslua [hunk 12]
```diff
@@ -0,0 +1,446 @@
...
+
+  local playerCardData
+  local locationData
+
+  local TokenManager = { }
+  local internal = { }
+
+  -- Spawns tokens for the card.  This function is built to just throw a card at it and let it do
+  -- the work.  It will check to see if the card has already spawned, find appropriate data from
+  -- either the uses metadata or the Data Helper, and spawn the tokens.
```
[the remark this answers is not in the document]
Comment (Buhallin) [reply src/core/token/TokenManager.ttslua:128]: Updated the comment, which was slightly outdated - original plan was to have this method handle it, but that only held while we had "everything spawns on every sensitive spot".  Once we added position-limited spawning, this method couldn't (or at least shouldn't) know about that and we had to do the check outside.

File: src/core/token/TokenManager.ttslua [hunk 13]
```diff
@@ -0,0 +1,446 @@
+do
+  local tokenSpawnTracker = require("core/token/TokenSpawnTrackerApi")
+  local playAreaApi = require("core/PlayAreaApi")
+
+  local PLAYER_CARD_TOKEN_OFFSETS = {
+    [1] = {
+      Vector(0, 8, -0.2)
+    },
```
[the remark this answers is not in the document]
Comment (Buhallin) [reply src/core/token/TokenManager.ttslua:8]: I've been thinking about that.  For now I wanted to minimize an already-large change so just kept the positions.

For the future... the manual here actually gives some nice specific layouts (like 3 spawning in a triangle) that would be impossible to replicate with a generated set of positions.  I'm definitely planning on pretty-ing up the clue positions, but for this I think we may be better leaving it.

File: src/core/token/TokenManager.ttslua [hunk 14]
```diff
@@ -0,0 +1,446 @@
...
+    local sourceBag = getObjectFromGUID(SOURCE_BAG_GUIDS[tokenType])
+    if sourceBag == nil then
+      error("No token source for '" .. tokenType .. "'")
+      return
+    end
+    -- All the source bags are infinite, so we just grab the first object
+    local tokenTemplate = sourceBag.getData().ContainedObjects[1]
+    -- Take ONLY the Y-value for rotation, so we don't flip the token coming out of the bag
+    local tokenRotation  = rotation or { x = 0, y = 270, z = 0 }
+    tokenTemplate.Transform.rotY = tokenRotation.y
```
[the remark this answers is not in the document]
Comment (Buhallin) [reply src/core/token/TokenManager.ttslua:249]: I don't mind additional variables :)  At least where they add clarity, which IMHO this does.  Especially combining the results of the Lua or weirdness, pulling a single value out of the separate result is much easier to read.

File: src/core/token/TokenManager.ttslua [hunk 15]
```diff
@@ -0,0 +1,446 @@
...
+  end
+
+  -- Checks to see if the given card has location data in the DataHelper
+  ---@param card Object Card to check for data
+  ---@return Boolean True if this card has data in the helper, false otherwise
+  TokenManager.hasLocationData = function(card)
+    return internal.getLocationData(card) ~= nil
+  end
+
+  -- Copies the data from the DataHelper.  Will only happen once.
```
[the remark this answers is not in the document]
Comment (Buhallin) [reply src/core/token/TokenManager.ttslua:290]: Copying the data was the pattern used by the previous implementation, so I just kept it.

It does simplify access - since the Customs add data to these tables, we'd have to either mutate the base DataHelper (bad) or keep a base list and an added list separately, which would make the lookups harder.

File: src/core/Global.ttslua (hunk 8 above)
[the remark this answers is not in the document]
Comment (Chr1Z93) [reply src/core/Global.ttslua:598]: What speaks against updating the campaigns in seth's repo?

File: src/core/MythosArea.ttslua (hunk 10 above)
[the remark this answers is not in the document]
Comment (Chr1Z93) [reply src/core/MythosArea.ttslua:11]: I don't mind where the values end up, but they should be in one place imo.
The "global" variables in the global object are still scoped to the global object and can only be accessed via `getVar` and `setVar` (not by accidentally reusing the names).

File: src/core/token/TokenManager.ttslua (hunk 13 above)
[the remark this answers is not in the document]
Comment (Chr1Z93) [reply src/core/token/TokenManager.ttslua:8]: It's fine to keep it as is, but of course this could also be generated automatically be checking if `i` is odd or even.

File: src/core/token/TokenManager.ttslua (hunk 14 above)
[the remark this answers is not in the document]
Comment (Chr1Z93) [reply src/core/token/TokenManager.ttslua:250]: Combining this into `tokenTemplate.Transform.rotY = rotation.y or 270` is more readable to me tbh

File: src/core/Global.ttslua (hunk 8 above)
[the remark this answers is not in the document]
Comment (Buhallin) [reply src/core/Global.ttslua:598]: It's a massive enough change that I'd like to confine it to the beta test until we do the full release.  With it touching something like 100 scenarios I don't think we can expect to test it thoroughly on our own

File: src/core/MythosArea.ttslua (hunk 10 above)
[the remark this answers is not in the document]
Comment (Buhallin) [reply src/core/MythosArea.ttslua:11]: Right, but what I mean is that the values in Global are table-space, not relative to the panel.  So if we were to move the panel as part of a table reorganization, those values would break (or at least need to be updated).

[the remark this answers is not in the document]
Comment (Buhallin) [reply src/core/MythosArea.ttslua:11]: And FWIW I agree that they should be in a single place.  But I don't want to perpetuate the use of getVar(), which is bad for all sorts of reasons, and refactoring the current encounter draw out of Global is way beyond what we have time for with the release target.

If you feel super strong about it I'll reference the vars in Global because it's the only reasonable option, but that's just replacing one bit of bad code with another and isn't improving anything.  Will leave this for you to resolve or not, let me know.

File: src/core/token/TokenManager.ttslua [hunk 16]
```diff
@@ -0,0 +1,447 @@
...
+    }
+  }
+
+  local SOURCE_BAG_GUIDS = {
+    damage = "480bda",
+    horror = "c3ecf4",
+    resource = "9fadf9",
+    doom = "47ffc3",
+    clue = "31fa39",
+  }
```
[the remark this answers is not in the document]
Comment (Buhallin) [reply src/core/token/TokenManager.ttslua:116]: Yeah, I'd considered that too...  probably should have a copy, or more likely convert it to a single standard bag as a source.  Will update tonight.

File: src/core/Global.ttslua (hunk 8 above)
[the remark this answers is not in the document]
Comment (Chr1Z93) [reply src/core/Global.ttslua:598]: Alright 👍

File: src/core/MythosArea.ttslua (hunk 10 above)
[the remark this answers is not in the document]
Comment (Chr1Z93) [reply src/core/MythosArea.ttslua:11]: Let's just keep this as is and maybe move variables later if needed

File: src/core/token/TokenManager.ttslua (hunk 16 above)
[the remark this answers is not in the document]
Comment (Buhallin) [reply src/core/token/TokenManager.ttslua:116]: Done - added a bag with copies of the tokens, keyed using GMNotes.

File: src/core/token/TokenManager.ttslua (hunk 14 above)
[the remark this answers is not in the document]
Comment (Buhallin) [reply src/core/token/TokenManager.ttslua:250]: With the templates this ended up more of a copy (to keep from changing the values on the template), so I think this is OBE

File: src/core/token/TokenManager.ttslua [hunk 17]
```diff
@@ -107,6 +107,9 @@ do
     }
   }
 
+  local TOKEN_SOURCE_GUID = "124381"
+  local tokenTemplates
+
   local SOURCE_BAG_GUIDS = {
```
Comment (Chr1Z93) [inline src/core/token/TokenManager.ttslua:113]: Source bag guids should no longer be needed, right?

**Review by Chr1Z93 — APPROVED** [review]
